# 실습 1. Excel 파일 불러오기
## 해야 할 일

pandas를 이용해 Excel 파일을 DataFrame으로 불러옵니다.

먼저 코드부터 작성하기보다 인공지능에게 현재 상황과 원하는 작업을 설명해 봅니다.

 


In [1]:
import pandas as pd

df = pd.read_excel(r'C:\dev\ai-data-analysis\data\raw\교보문고_온라인_베스트셀러_상품리스트.xlsx')

df.head()

c:\dev\ai-data-analysis\.venv\Lib\site-packages\openpyxl\styles\stylesheet.py:237: UserWarning: Workbook contains no default style, apply openpyxl's default
  warn("Workbook contains no default style, apply openpyxl's default")


,순위,상품코드,판매상품 ID,상품명,정가,판매가,할인율,적립율,적립예정포인트,인물,출판사,발행(출시)일자,분야
0,1,9791199489561,S000220308313,"세네카, 오늘을 빼앗기고 있는 당신에게","18,000","16,200",10%,5%,900,세네카,논픽션,20260701.0,인문
1,2,9791198742179,S000221143299,홍정기의 장수근육 혁명,"19,800","17,820",10%,5%,990,홍정기,깸,20260831.0,건강
2,3,9788937460586,S000000620195,싯다르타,"8,000","7,200",10%,5%,400,헤르만 헤세,민음사,20020120.0,소설
3,4,9791124038826,S000221075112,머니 트렌드 2027,"27,500","24,750",10%,5%,"1,370",정태익 외,북모먼트,20260916.0,경제/경영
4,5,9791175488854,S000221248222,흔한남매 23,"16,800","15,120",10%,5%,840,흔한남매,미래엔아이세움,20260916.0,어린이(초등)


## pandas로 Excel 파일 읽기

`교보문고_온라인_베스트셀러_상품리스트.xlsx` 파일을 pandas의 `read_excel()`로 불러와 `df`에 저장하고, `df.head()`로 상위 5행을 확인했다.

데이터는 상품코드, 판매상품ID, 상품명, 정가, 판매가, 할인율, 적립율, 적립예정포인트, 인물(저자), 출판사, 발행(출시)일자, 분야 등의 컬럼으로 구성되어 있으며, 온라인 베스트셀러 도서 목록임을 확인할 수 있었다.

경로에 한글 폴더명과 백슬래시가 섞여 있어 `OSError`가 발생했는데, 문자열 앞에 `r`을 붙여 raw string으로 처리하니 정상적으로 실행되었다.

### 1위 상품 분석

상위 5행 중 1위는 세네카의 《오늘을 빼앗기고 있는 당신에게》로, 정가 18,000원에서 10% 할인된 16,200원에 판매되고 있으며, 구매 시 5%(900포인트)를 적립받을 수 있다. 2026년 7월 1일에 발행된 인문 분야 도서로, 최근 발행된 책임에도 베스트셀러 상위에 올라 있다는 점이 눈에 띈다.

### 실습 2. 데이터의 크기와 컬럼 확인하기
Excel 파일을 불러왔으면 바로 전처리를 시작하지 않습니다.

먼저 어떤 데이터인지 확인합니다.


In [2]:
print(df.shape)  # 1. 전체 행과 컬럼 개수
print(df.columns)  # 2. 컬럼 이름 목록
print(df.dtypes)  # 3. 각 컬럼의 자료형
print(df.isna().sum())  # 4. 각 컬럼의 결측치 개수
df.head()  # 5. 앞의 5행

(989, 13)
Index(['순위', '상품코드', '판매상품 ID', '상품명', '정가', '판매가', '할인율', '적립율', '적립예정포인트',
       '인물', '출판사', '발행(출시)일자', '분야'],
      dtype='str')
순위            int64
상품코드            str
판매상품 ID         str
상품명             str
정가              str
판매가             str
할인율             str
적립율             str
적립예정포인트         str
인물              str
출판사             str
발행(출시)일자    float64
분야              str
dtype: object
순위           0
상품코드         0
판매상품 ID      0
상품명          0
정가           0
판매가          0
할인율          0
적립율          0
적립예정포인트      0
인물           3
출판사          3
발행(출시)일자     3
분야          62
dtype: int64


,순위,상품코드,판매상품 ID,상품명,정가,판매가,할인율,적립율,적립예정포인트,인물,출판사,발행(출시)일자,분야
0,1,9791199489561,S000220308313,"세네카, 오늘을 빼앗기고 있는 당신에게","18,000","16,200",10%,5%,900,세네카,논픽션,20260701.0,인문
1,2,9791198742179,S000221143299,홍정기의 장수근육 혁명,"19,800","17,820",10%,5%,990,홍정기,깸,20260831.0,건강
2,3,9788937460586,S000000620195,싯다르타,"8,000","7,200",10%,5%,400,헤르만 헤세,민음사,20020120.0,소설
3,4,9791124038826,S000221075112,머니 트렌드 2027,"27,500","24,750",10%,5%,"1,370",정태익 외,북모먼트,20260916.0,경제/경영
4,5,9791175488854,S000221248222,흔한남매 23,"16,800","15,120",10%,5%,840,흔한남매,미래엔아이세움,20260916.0,어린이(초등)


## 데이터 기본 구조 확인

- 전체 데이터는 989행, 13개 컬럼으로 구성되어 있다.
- 컬럼 목록: 순위, 상품코드, 판매상품 ID, 상품명, 정가, 판매가, 할인율, 적립율, 적립예정포인트, 인물, 출판사, 발행(출시)일자, 분야

### 자료형 확인

- `순위`는 정수형(int64), `발행(출시)일자`는 실수형(float64)으로 되어 있다.
- `정가`, `판매가`, `할인율`, `적립율`, `적립예정포인트` 등 숫자로 보이는 컬럼들이 실제로는 문자형(str)으로 저장되어 있어, 이후 계산을 하려면 숫자형으로 변환하는 전처리가 필요하다.

### 결측치 확인

- `인물`, `출판사`, `발행(출시)일자`는 각각 3개의 결측치가 있다.
- `분야`는 62개로 결측치가 가장 많아, 분야별 분석 시 주의가 필요하다.
- 그 외 컬럼(순위, 상품코드, 상품명, 정가, 판매가 등)은 결측치가 없다.

In [3]:
df.isna().sum().sum()

np.int64(71)

## 실습 3. 분석에 필요한 컬럼만 선택하기

 

원본 데이터의 모든 컬럼이 이후 분석에 필요한 것은 아닙니다.

 

이번 과정에서는 다음 컬럼을 먼저 선택합니다.


In [4]:
# 원하는 컬럼만 리스트로 지정해서 새 DataFrame 만들기
cols = ['순위', '판매상품 ID', '상품명', '판매가', '인물', '출판사', '발행(출시)일자', '분야']
df_books = df[cols]
# 컬럼이 잘 선택되었는지 확인
print(df_books.columns)  # 컬럼 이름 목록 확인
df_books.head()  # 상위 5행으로 데이터 확인

Index(['순위', '판매상품 ID', '상품명', '판매가', '인물', '출판사', '발행(출시)일자', '분야'], dtype='str')


,순위,판매상품 ID,상품명,판매가,인물,출판사,발행(출시)일자,분야
0,1,S000220308313,"세네카, 오늘을 빼앗기고 있는 당신에게","16,200",세네카,논픽션,20260701.0,인문
1,2,S000221143299,홍정기의 장수근육 혁명,"17,820",홍정기,깸,20260831.0,건강
2,3,S000000620195,싯다르타,"7,200",헤르만 헤세,민음사,20020120.0,소설
3,4,S000221075112,머니 트렌드 2027,"24,750",정태익 외,북모먼트,20260916.0,경제/경영
4,5,S000221248222,흔한남매 23,"15,120",흔한남매,미래엔아이세움,20260916.0,어린이(초등)


## 분석에 필요한 컬럼만 선택하기

원본 `df`에서 분석에 필요한 8개 컬럼(순위, 판매상품 ID, 상품명, 판매가, 인물, 출판사, 발행(출시)일자, 분야)만 선택하여 `df_books`라는 새 DataFrame을 만들었다.

`df_books.columns`로 원하는 컬럼이 정확히 선택되었는지 확인했고, `df_books.head()`로 확인한 결과 순위 1위부터 5위까지의 상품명, 판매가, 저자, 출판사 등 핵심 정보만 깔끔하게 남은 것을 확인할 수 있었다.

불필요한 컬럼(상품코드, 정가, 할인율, 적립율, 적립예정포인트)을 제외함으로써 이후 분석 작업이 더 간단해질 것으로 예상된다.

## 실습 4. 컬럼 이름을 사용하기 쉽게 정리하기
원본 컬럼 이름은 사람이 읽기에는 좋지만 코드에서 반복해서 사용하기에는 조금 길 수 있습니다.

이번 실습에서는 다음처럼 간단하게 정리할 수 있습니다.


In [5]:
# 컬럼 이름 일부를 딕셔너리로 매핑해서 변경
df_books = df_books.rename(columns={
    '판매상품 ID': '판매상품ID',
    '인물': '저자',
    '발행(출시)일자': '발행일'
})

df_books.columns  # 변경된 컬럼 이름 확인

Index(['순위', '판매상품ID', '상품명', '판매가', '저자', '출판사', '발행일', '분야'], dtype='str')

## 컬럼 이름을 사용하기 쉽게 정리하기

`rename()`을 사용해 `df_books`의 컬럼 이름 중 일부를 더 짧고 간단한 이름으로 변경했다.

- 판매상품 ID -> 판매상품ID
- 인물 -> 저자
- 발행(출시)일자 -> 발행일

변경 후 `df_books.columns`로 확인한 결과, 지정한 3개 컬럼만 이름이 바뀌고 나머지 컬럼(순위, 상품명, 판매가, 출판사, 분야)은 그대로 유지된 것을 확인했다. 공백이나 괄호가 없는 이름으로 정리되어 앞으로 코드에서 컬럼을 다룰 때 더 편해질 것으로 보인다.

## 실습 5. 결측치 확인하기


결측치는 값이 비어 있는 데이터를 의미합니다.

데이터 분석에서는 결측치를 무조건 삭제하지 않습니다.

먼저 어디에 얼마나 있는지 확인하고, 분석 목적에 맞게 처리 방법을 결정합니다.


In [6]:
# 컬럼별 결측치 개수 확인
df_books.isna().sum()

순위         0
판매상품ID     0
상품명        0
판매가        0
저자         3
출판사        3
발행일        3
분야        62
dtype: int64

In [7]:
# 결측치 개수 세기
missing = df_books.isna().sum()

# 결측치가 1개 이상인 컬럼만 보기
missing[missing > 0]

저자      3
출판사     3
발행일     3
분야     62
dtype: int64

## 결측치 확인하기

`df_books.isna().sum()`으로 컬럼별 결측치 개수를 확인했다. 순위, 판매상품ID, 상품명, 판매가는 결측치가 없었고, 저자·출판사·발행일은 각각 3개, 분야는 62개의 결측치가 있었다.

결측치가 1개 이상인 컬럼만 `missing[missing > 0]`으로 걸러서 확인한 결과 저자, 출판사, 발행일, 분야 4개 컬럼에만 결측치가 존재함을 다시 한 번 확인했다. 특히 분야 컬럼은 결측치 비율이 눈에 띄게 높아, 이후 분석 시 별도의 처리 방법을 고민해야 할 것으로 보인다.

## 실습 6. 판매가를 숫자로 바꾸기

 

Excel에서 판매가가 다음과 같이 저장되어 있을 수 있습니다.


In [8]:
# 1. 쉼표(,)를 빈 문자열로 바꿔서 제거
df_books['판매가'] = df_books['판매가'].str.replace(',', '')

# 2. 문자열을 숫자형(정수)으로 변환
df_books['판매가'] = df_books['판매가'].astype(int)

# 3. 변환된 dtype 확인
df_books['판매가'].dtype

dtype('int64')

## 판매가를 숫자로 바꾸기

`판매가` 컬럼이 `'16,200'`처럼 쉼표가 포함된 문자열로 저장되어 있어 계산에 바로 사용할 수 없었다.

`str.replace(',', '')`로 쉼표를 제거한 뒤 `astype(int)`로 정수형으로 변환했고, 변환 후 dtype을 확인한 결과 `int64`로 정상 변환된 것을 확인했다.

이제 판매가 컬럼으로 합계, 평균, 정렬 등 숫자 기반 분석이 가능해졌다.

## 실습 7. 발행일을 날짜형으로 바꾸기

 

원본 데이터의 발행일은 다음처럼 보일 수 있습니다.

In [9]:
# 1. float 타입의 발행일을 정수형(Int64, 결측치 허용)으로 바꾼 뒤 문자열로 변환
df_books['발행일'] = df_books['발행일'].astype('Int64').astype(str)

# 2. YYYYMMDD 형식의 문자열을 날짜형으로 변환 (형식이 안 맞으면 NaT 처리)
df_books['발행일'] = pd.to_datetime(df_books['발행일'], format='%Y%m%d', errors='coerce')

# 3. 변환된 dtype과 앞의 5개 값 확인
print(df_books['발행일'].dtype)
df_books['발행일'].head()

datetime64[us]


0   2026-07-01
1   2026-08-31
2   2002-01-20
3   2026-09-16
4   2026-09-16
Name: 발행일, dtype: datetime64[us]

## 발행일을 날짜형으로 바꾸기

`발행일` 컬럼이 `20260701.0`처럼 실수형(float)으로 저장되어 있어 날짜로 다루기 어려웠다.

먼저 `Int64`로 변환해 결측치를 유지한 채 정수로 바꾸고, 다시 문자열로 변환한 뒤 `pd.to_datetime()`에 `format='%Y%m%d'`를 지정해 날짜형으로 변환했다. `errors='coerce'` 옵션 덕분에 형식이 맞지 않거나 결측치인 값은 오류 없이 `NaT`로 처리되었다.

변환 결과 dtype이 `datetime64[us]`로 바뀐 것을 확인했고, 앞의 5개 값을 보면 2026-07-01, 2026-08-31, 2002-01-20, 2026-09-16, 2026-09-16처럼 실제 날짜 형태로 잘 표시되었다.


##  실습 8. 문자열 앞뒤 공백 정리하기
상품명, 저자, 출판사, 분야는 문자열 데이터입니다.

문자열 앞뒤에 불필요한 공백이 있으면 같은 값도 서로 다른 데이터처럼 처리될 수 있습니다.

예를 들어 다음 두 값은 사람이 보기에는 같지만 문자열로는 다릅니다.


In [10]:
# 공백을 제거할 컬럼 목록
cols_to_strip = ['상품명', '저자', '출판사', '분야']

# 반복문으로 각 컬럼에 str.strip() 적용
for col in cols_to_strip:
    df_books[col] = df_books[col].str.strip()

# 처리 후 값 확인 (컬럼 4개를 한 번에 5행씩)
df_books[cols_to_strip].head()

,상품명,저자,출판사,분야
0,"세네카, 오늘을 빼앗기고 있는 당신에게",세네카,논픽션,인문
1,홍정기의 장수근육 혁명,홍정기,깸,건강
2,싯다르타,헤르만 헤세,민음사,소설
3,머니 트렌드 2027,정태익 외,북모먼트,경제/경영
4,흔한남매 23,흔한남매,미래엔아이세움,어린이(초등)


## 문자열 앞뒤 공백 제거하기

`상품명`, `저자`, `출판사`, `분야` 컬럼에 눈에 보이지 않는 앞뒤 공백이 섞여 있으면, 실제로는 같은 값인데도 서로 다른 문자열로 인식되어 그룹핑이나 집계 시 오류가 생길 수 있다.

반복문으로 4개 컬럼에 각각 `str.strip()`을 적용해 앞뒤 공백을 제거했다. 처리 후 값을 확인한 결과 데이터 자체는 그대로 유지되면서 불필요한 공백만 정리된 것을 확인했다.

이 작업으로 이후 분야별 집계나 저자·출판사 기준 그룹화 작업에서 공백 때문에 생기는 오류를 예방할 수 있다.

## 실습 9. 중복 데이터 확인하기

데이터에서 중복을 발견했다고 해서 바로 삭제하면 안 됩니다.

 
먼저 무엇을 기준으로 중복인지 생각해야 합니다.


예를 들어 같은 상품명이 여러 번 등장할 수 있습니다.



하지만 같은 제목이라도 다른 판본, 다른 상품, 개정판 등이 있을 수 있습니다.

 
따라서 이번 데이터에서는 상품명만 보고 중복 행을 바로 삭제하지 않습니다.

 
우선 상품을 구분하는 판매상품ID를 확인합니다.


In [11]:
# 1. 전체 행이 완전히 같은 중복 행 개수
df_books.duplicated().sum()

np.int64(0)

In [12]:
# 2. 판매상품ID가 중복된 행 개수
df_books.duplicated(subset='판매상품ID').sum()

np.int64(0)

In [13]:
# 3. 상품명이 중복된 행 개수
df_books.duplicated(subset='상품명').sum()

np.int64(37)

## 중복 데이터 확인하기

`duplicated()`를 이용해 세 가지 기준으로 중복 여부를 확인했다.

- 전체 행이 완전히 같은 중복: 0건
- 판매상품ID 기준 중복: 0건
- 상품명 기준 중복: 37건

판매상품ID는 상품을 구분하는 고유 식별자이므로 중복이 없는 것이 정상이며, 실제로도 중복이 확인되지 않았다. 반면 상품명 기준으로는 37건의 중복이 발견되었는데, 이는 같은 책이 다른 판매상품ID(예: 다른 판본, 세트 구성 등)로 여러 번 등록되었을 가능성을 의미한다. 상품을 구분할 때는 상품명이 아니라 판매상품ID를 기준으로 삼아야 한다는 것을 알 수 있다.

##  실습 10. 전처리 결과 검증하기

지금까지 여러 전처리를 수행했습니다.

마지막에는 반드시 데이터가 의도한 형태가 되었는지 다시 확인합니다.


In [14]:

print(df_books.shape)  # 전체 행과 컬럼 개수
print(df_books.columns)  # 컬럼 이름 목록
print(df_books.dtypes)  # 각 컬럼의 자료형
print(df_books.isna().sum())  # 컬럼별 결측치 개수
print(df_books['판매가'].head())  # 판매가 앞의 5개 값
print(df_books['발행일'].head())  # 발행일 앞의 5개 값
print(df_books.duplicated(subset='판매상품ID').sum())  # 판매상품ID 중복 개수
df_books.head()  # 전체 데이터 앞의 5행


(989, 8)
Index(['순위', '판매상품ID', '상품명', '판매가', '저자', '출판사', '발행일', '분야'], dtype='str')
순위                 int64
판매상품ID               str
상품명                  str
판매가                int64
저자                   str
출판사                  str
발행일       datetime64[us]
분야                   str
dtype: object
순위         0
판매상품ID     0
상품명        0
판매가        0
저자         3
출판사        3
발행일        3
분야        62
dtype: int64
0    16200
1    17820
2     7200
3    24750
4    15120
Name: 판매가, dtype: int64
0   2026-07-01
1   2026-08-31
2   2002-01-20
3   2026-09-16
4   2026-09-16
Name: 발행일, dtype: datetime64[us]
0


,순위,판매상품ID,상품명,판매가,저자,출판사,발행일,분야
0,1,S000220308313,"세네카, 오늘을 빼앗기고 있는 당신에게",16200,세네카,논픽션,2026-07-01,인문
1,2,S000221143299,홍정기의 장수근육 혁명,17820,홍정기,깸,2026-08-31,건강
2,3,S000000620195,싯다르타,7200,헤르만 헤세,민음사,2002-01-20,소설
3,4,S000221075112,머니 트렌드 2027,24750,정태익 외,북모먼트,2026-09-16,경제/경영
4,5,S000221248222,흔한남매 23,15120,흔한남매,미래엔아이세움,2026-09-16,어린이(초등)


## 전처리 결과 최종 확인

전처리를 마친 `df_books`는 989행, 8개 컬럼(순위, 판매상품ID, 상품명, 판매가, 저자, 출판사, 발행일, 분야)으로 구성되어 있다.

자료형을 확인한 결과 `판매가`는 정수형(int64), `발행일`은 날짜형(datetime64[us])으로 목표한 대로 변환되어 있었다. 판매가 앞의 5개 값(16200, 17820, 7200, 24750, 15120)과 발행일 값(2026-09-16 등)도 각각 숫자와 날짜 형태로 정상 출력되었다.

결측치는 저자·출판사·발행일이 각각 3개, 분야가 62개로 이전 확인 결과와 동일했고, 판매상품ID 기준 중복은 0건으로 상품 식별자가 고유하게 유지되고 있음을 다시 확인했다.

## 실습 11. 전처리 데이터 저장하기

다음 Chapter에서는 이 데이터를 다시 사용합니다.

따라서 전처리한 DataFrame을 CSV 파일로 저장합니다.

추천 파일명은 다음과 같습니다.


In [15]:
df_books.to_csv('book_bestseller_clean.csv', index=False, encoding='utf-8-sig')

## 전처리 결과 CSV로 저장하기

전처리가 끝난 `df_books`를 `to_csv()`를 이용해 `book_bestseller_clean.csv` 파일로 저장했다. `index=False` 옵션으로 불필요한 인덱스 컬럼이 파일에 포함되지 않도록 했고, `encoding='utf-8-sig'`를 사용해 Windows 엑셀에서 열어도 한글이 깨지지 않도록 처리했다.

파일 탐색기에서 `book_bestseller_clean.csv` 파일이 정상적으로 생성된 것을 확인했다.

## 12. 인공지능을 이용해 결과 Markdown 작성하기

 

실습을 완료하면 Notebook에 이번 작업을 설명하는 Markdown을 작성합니다.

 

여기서도 인공지능을 사용할 수 있습니다.

 

단, 인공지능이 실행 결과를 추측하게 하면 안 됩니다.

 

Notebook에서 실제로 확인한 숫자를 프롬프트에 넣어야 합니다.

## 정리

**1. 무엇을 확인했는지**

원본 데이터는 989행, 13개 컬럼으로 이루어져 있었고, `shape`, `columns`, `dtypes`, `isna().sum()`을 이용해 컬럼 구성과 자료형, 결측치 현황을 확인했다. 그 결과 저자 3개, 분야 62개의 결측치가 있었고, 판매가·정가 등 숫자로 보이는 컬럼이 실제로는 문자열로 저장되어 있는 것을 확인했다.

**2. 어떤 전처리를 했는지**

분석에 필요한 8개 컬럼(순위, 판매상품ID, 상품명, 판매가, 저자, 출판사, 발행일, 분야)만 선택해 `df_books`를 만들고, 컬럼 이름을 더 쓰기 쉽게 정리했다. 판매가는 쉼표를 제거한 뒤 숫자형으로 바꾸었고, 발행일은 날짜형으로 변환했다. 문자열 컬럼의 앞뒤 공백도 제거했고, 판매상품ID를 기준으로 중복 여부를 확인한 결과 중복은 0건이었다.

**3. 다음 분석을 위해 어떤 상태가 되었는지**

판매가는 숫자 계산이 가능한 상태, 발행일은 날짜 계산이 가능한 상태가 되어 이후 평균 가격 계산이나 발행 연도별 분석 등을 바로 진행할 수 있는 상태가 되었다.

## 13. 이번 Chapter에서 꼭 기억할 것

 

데이터 전처리에서 중요한 것은 많은 코드를 작성하는 것이 아닙니다.

 

먼저 데이터를 확인하고, 문제가 무엇인지 판단한 뒤 필요한 처리를 하는 것이 중요합니다.

 

특히 다음 내용을 기억합니다.

## 14. 간단한 확인 문제

**질문 1. 왜 모든 원본 컬럼을 사용하지 않고 필요한 컬럼만 선택했나요?**

원본 데이터에는 정가, 할인율, 적립율, 적립예정포인트처럼 이번 분석 목적과 크게 관련 없는 컬럼도 섞여 있었다. 불필요한 컬럼까지 모두 들고 있으면 데이터를 다루기 복잡해지고, 실수로 관련 없는 값을 분석에 끼워 넣을 위험도 커진다. 필요한 컬럼만 골라두면 이후 코드가 더 단순해지고 실수도 줄어든다.

**질문 2. 판매가가 문자열이라면 어떤 문제가 발생할 수 있나요?**

문자열 상태로는 합계, 평균, 정렬 같은 숫자 계산이 불가능하거나 엉뚱한 결과가 나올 수 있다. 예를 들어 `'16,200'`처럼 쉼표가 들어 있으면 숫자로 인식되지 않아 계산 시 오류가 나거나, 문자열 기준으로 정렬되어 크기 순서가 실제 숫자 순서와 다르게 나올 수 있다.

**질문 3. 결측치를 확인한 뒤 바로 모든 행을 삭제하면 안 되는 이유는 무엇인가요?**

결측치가 있는 컬럼과 개수는 컬럼마다 다르다(예: 분야는 62개, 저자는 3개). 결측치가 있는 행을 무조건 다 지우면 분야 결측 62건 때문에 저자·출판사 정보는 멀쩡한 데이터까지 함께 사라져서 데이터 손실이 커진다. 어떤 컬럼의 결측치가 분석에 실제로 중요한지 먼저 판단한 뒤, 필요한 경우에만 제거하거나 다른 방식으로 처리해야 한다.

**질문 4. 같은 상품명이 두 번 등장했다고 해서 바로 중복 데이터라고 판단하기 어려운 이유는 무엇인가요?**

같은 책이라도 판형, 세트 구성, 개정판 등으로 인해 상품명은 같지만 실제로는 서로 다른 상품(판매상품ID가 다름)인 경우가 있을 수 있다. 실제로 상품명 기준 중복은 37건이었지만 판매상품ID 기준 중복은 0건이었던 것처럼, 상품명만으로 중복을 판단하면 서로 다른 상품을 같은 것으로 잘못 처리할 수 있다.

**질문 5. 인공지능이 작성한 코드가 오류 없이 실행되었다면 그것만으로 분석이 끝난 것일까요?**

아니다. 코드가 오류 없이 실행되는 것과 결과가 의도한 대로 맞게 나온 것은 다른 문제다. 변환된 dtype이 예상한 자료형인지, 결측치나 중복 개수가 합리적인지 직접 눈으로 확인하고 그 의미를 이해해야 분석이 제대로 끝났다고 할 수 있다.

## 15. 제출 전 확인

Notebook을 처음부터 마지막 셀까지 다시 실행한 뒤 아래 항목을 확인했다.

- [x] Excel 파일이 정상적으로 로드된다.
- [x] 데이터 크기와 컬럼을 확인했다.
- [x] 필요한 컬럼만 선택했다.
- [x] 컬럼 이름을 정리했다.
- [x] 결측치를 확인하고 처리했다.
- [x] 판매가를 숫자형으로 변환했다.
- [x] 발행일을 날짜형으로 변환했다.
- [x] 문자열 앞뒤 공백을 정리했다.
- [x] 중복 데이터를 확인했다.
- [x] 최종 데이터를 다시 검증했다.
- [x] book_bestseller_clean.csv를 저장했다.
- [x] 실제 실행 결과를 바탕으로 Markdown 설명을 작성했다.
- [x] Notebook 전체를 다시 실행했을 때 오류가 없다.